### ISSUE TITLE
Glassfish: MOC Cloud DV Glassfish crashing

### ISSUE THREAD

**Issue by ferrys** [IQSS/dataverse#4416]
At the MOC we're having a problem with Glassfish crashing very sporadically in our to-be production environment. We're running Glassfish 4.1, PostgreSQL 9.2.23 and Java openjdk version 1.8.0_151 on a Centos image with 48GB of RAM.

Also, these are our current JVM options:
```
[root@dataverse domain1]# /usr/local/glassfish4/bin/asadmin list-jvm-options
-server
-Djava.awt.headless=true
-Djdk.corba.allowOutputStreamSubclass=true
-Djavax.xml.accessExternalSchema=all
-Djavax.management.builder.initial=com.sun.enterprise.v3.admin.AppServerMBeanServerBuilder
-XX:+UnlockDiagnosticVMOptions
-Djava.endorsed.dirs=${com.sun.aas.installRoot}/modules/endorsed${path.separator}${com.sun.aas.installRoot}/lib/endorsed
-Djava.security.policy=${com.sun.aas.instanceRoot}/config/server.policy
-Djava.security.auth.login.config=${com.sun.aas.instanceRoot}/config/login.conf
-Dcom.sun.enterprise.security.httpsOutboundKeyAlias=s1as
-Djavax.net.ssl.keyStore=${com.sun.aas.instanceRoot}/config/keystore.jks
-Djavax.net.ssl.trustStore=${com.sun.aas.instanceRoot}/config/cacerts.jks
-Djava.ext.dirs=${com.sun.aas.javaRoot}/lib/ext${path.separator}${com.sun.aas.javaRoot}/jre/lib/ext${path.separator}${com.sun.aas.instanceRoot}/lib/ext
-Djdbc.drivers=org.apache.derby.jdbc.ClientDriver
-DANTLR_USE_DIRECT_CLASS_LOADING=true
-Dcom.sun.enterprise.config.config_environment_factory_class=com.sun.enterprise.config.serverbeans.AppserverConfigEnvironmentFactory
-Dorg.glassfish.additionalOSGiBundlesToStart=org.apache.felix.shell,org.apache.felix.gogo.runtime,org.apache.felix.gogo.shell,org.apache.felix.gogo.command,org.apache.felix.shell.remote,org.apache.felix.fileinstall
-Dosgi.shell.telnet.port=6666
-Dosgi.shell.telnet.maxconn=1
-Dosgi.shell.telnet.ip=127.0.0.1
-Dgosh.args=--nointeractive
-Dfelix.fileinstall.dir=${com.sun.aas.installRoot}/modules/autostart/
-Dfelix.fileinstall.poll=5000
-Dfelix.fileinstall.log.level=2
-Dfelix.fileinstall.bundles.new.start=true
-Dfelix.fileinstall.bundles.startTransient=true
-Dfelix.fileinstall.disableConfigSave=false
-XX:NewRatio=2
-Dcom.ctc.wstx.returnNullForDefaultNamespace=true
-XX:MaxPermSize=512m
-XX:PermSize=256m
-Xmx24g
-Xms24g
-Ddataverse.files.directory=/usr/local/glassfish4/glassfish/domains/domain1/files
-Ddataverse.rserve.host=localhost
-Ddataverse.rserve.port=6311
-Ddataverse.rserve.user=rserve
-Ddataverse.auth.password-reset-timeout-in-minutes=60
-Djavax.xml.parsers.SAXParserFactory=com.sun.org.apache.xerces.internal.jaxp.SAXParserFactoryImpl
-Ddoi.baseurlstring=https://ezid.cdlib.org
-Ddataverse.timerServer=true
-Ddataverse.files.storage-driver-id=swift
-Dawt.toolkit=sun.awt.HToolkit
-Ddataverse.fqdn=dataverse.massopen.cloud
-Ddataverse.file.swift-folder-path-separator=-
Command list-jvm-options executed successfully.
```

One time glassfish crashed it was accompanied by this error in the server.log:
```
Caused by: javax.persistence.PersistenceException: Exception [EclipseLink-4002] (Eclipse Persistence Services - 2.5.2.v20140319-9ad6abd): org.eclipse.persistence.exceptions.DatabaseException
Internal Exception: org.postgresql.util.PSQLException: This connection has been closed.
Error Code: 0
Call: SELECT NAME, CONTENT FROM SETTING
Query: ReadAllQuery(name="Setting.findAll" referenceClass=Setting sql="SELECT NAME, CONTENT FROM SETTING")
        at org.eclipse.persistence.internal.jpa.QueryImpl.getDetailedException(QueryImpl.java:378)
        at org.eclipse.persistence.internal.jpa.QueryImpl.executeReadQuery(QueryImpl.java:260)
        at org.eclipse.persistence.internal.jpa.QueryImpl.getResultList(QueryImpl.java:469)
        at edu.harvard.iq.dataverse.settings.SettingsServiceBean.listAll(SettingsServiceBean.java:486)
        at sun.reflect.GeneratedMethodAccessor474.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.glassfish.ejb.security.application.EJBSecurityManager.runMethod(EJBSecurityManager.java:1081)
        at org.glassfish.ejb.security.application.EJBSecurityManager.invoke(EJBSecurityManager.java:1153)
        at com.sun.ejb.containers.BaseContainer.invokeBeanMethod(BaseContainer.java:4786)
        at com.sun.ejb.EjbInvocation.invokeBeanMethod(EjbInvocation.java:656)
        at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
        at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:608)
        at org.jboss.weld.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:64)
        at org.jboss.weld.ejb.SessionBeanInterceptor.aroundInvoke(SessionBeanInterceptor.java:52)
        at sun.reflect.GeneratedMethodAccessor104.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:883)
        at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
        at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:608)
        at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.doCall(SystemInterceptorProxy.java:163)
        at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.aroundInvoke(SystemInterceptorProxy.java:140)
        at sun.reflect.GeneratedMethodAccessor105.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:883)
        at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
        at com.sun.ejb.containers.interceptors.InterceptorManager.intercept(InterceptorManager.java:369)
        at com.sun.ejb.containers.BaseContainer.__intercept(BaseContainer.java:4758)
        at com.sun.ejb.containers.BaseContainer.intercept(BaseContainer.java:4746)
        at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:212)
        ... 87 more
Caused by: Exception [EclipseLink-4002] (Eclipse Persistence Services - 2.5.2.v20140319-9ad6abd): org.eclipse.persistence.exceptions.DatabaseException
Internal Exception: org.postgresql.util.PSQLException: This connection has been closed.
```

But otherwise we have not been able to recreate it and it has also happened without this error. I will update this issue if/when it crashes again with the error. Any hints/guidance would be helpful to myself and @radonm!

**Comment by pameyer** [comment]
@ferrys it looks like a postgres connection failure to me.  Is there anything informative around that time in the postgres logs, and could you let us know which postgres JDBC version you're using.


FYI - when you go to production; some of those JVM options will be things that you won't want to post publicly.

**Comment by pameyer** [comment]
During some discussion with @ferrys , I'd mentioned that this looks to me like it might be related to a postgresql failure, and @ferrys mentioned storage problems.  If the storage failed, that could trigger a postgresql failure, and in turn result in glassfish loosing the connection to postgres.  Not sure if it's the cause in this case or not though.

**Comment by ferrys** [comment]
@pameyer Yea, I'm thinking it is definitely possible that this is what caused the crash when we got this error. 

We're using postgresql-9.1-902.jdbc4.jar & specifically CentOS Linux release 7.4.1708 (Core). Also, I'm not sure if our postgres is logging or if we just lost the logs when we had issues.

**Comment by pdurbin** [comment]
@radonm we recently merged a pull request to recommend PostgreSQL 9.6 if you'd like to check it out: #4721 . I'm not sure if it will help or not. Is Glassfish still crashing?

**Comment by djbrooke** [comment]
I'm going to close this. @radonm, please comment here and let us know if this is still an issue you're trying to resolve and, if so, how we can help with it.
